Added Tests to Schulze Functions in Core Package.

CheatCodeSam opened the pull request:
Added a few starter unit test for the Schulze functions. 
I added an init file in the test folder so it would be easier to import the elekto module.
I added pytests and pandas into the requirements.txt so that they would be installed with pip install.
I gave the Schulze functions type hints so they're a little easier to work around with, and write tests around. 

Let me know if this was what you were looking for, I'm happy to keep testing the other modules of this application.

jberkus commented:
Why Pandas?   Asking because that imports a LOT of modules.

jberkus commented:
Otherwise, this looks great, lemme try it out.

CheatCodeSam commented:
It's was already being used within `test/test_main.py`, and it looks like `elekto.core.Election.from_csv` expects a Dataframe object. I just went ahead and made sure if it was being installed with the rest of the requirements.txt. 

In all honesty it probably wouldn't be too difficult to remove the need for the dependency entirely.

jberkus commented:
@CheatCodeSam sorry for delay here, as expected your adding these tests exposed some unrelated problems that prevent testing the PR, specifically;

1. we need a way to run tests that don't require configuring a database, probably by loading SQLite.
2. also, the "no database found" Exception itself has a bug 

I'll work on these starting with (2), of course your ideas welcome.

CheatCodeSam commented:
Awesome, thank you for getting back to me on that. I've been poking around a few of the internals some more and I might have found some stuff that we could leverage for that first one. I'll append it to the PR and keep you in the loop.

jberkus commented:
Getting an error now:

```E   ModuleNotFoundError: No module named 'flask_sqlalchemy'```

jberkus commented:
Also, before you go to far with SQLalchemy testing: there are some really wonky bits with the database schema, particularly the "encrypted join" between voters and ballots.

jberkus commented:
Incidentally, I'd actually cleared the original changes (the replacement of the condorcet tests); we could merge that if you want to just PR in those commits.

CheatCodeSam commented:
We could merge those for now, I feel like it would nice to get something new into the repo since it hasn't been touched for a few years.

I'm working on getting the Flask App to run different configs depending if it's in a test env or not, then we can decide which database the tests run against.

Did you want these unit tests to run against a database? We could just try to mock everything out, but there would be a lot of stuff that needs to be snubbed even just to get a mock user going.

If we were running against a database, did you really want to use SQLite over something else. I'm thinking if this thing runs on Postgres/MySQL in Production, it should probably just test against that anyways.

Sorry for dumb questions, I'm really just trying to feel this thing out. I never realized how complicated testing Flask is lol

CheatCodeSam commented:
> Getting an error now:
> 
> `E RuntimeError: Either 'SQLALCHEMY_DATABASE_URI' or 'SQLALCHEMY_BINDS' must be set.`

Reverted this, I was toying around with the flask_sqlalchemy package.

CheatCodeSam commented:
I just added a commit you might be interested in. I changed to config to include a test database the tests run against. I also wrote a passing test that checks for an authenticated user.

jberkus commented:
Sam:

Some modules need to run against a database.  Also, a lot of classes just won't load unless the database is initialized; we'd need to put testing bypass logic in those to avoid it.  Sorry, that's a side effect of making Elekto a fully automated container.

Elekto is supposed to be fully functional against SQLite, so I'm happy to do unit tests against SQLite.  We can do E2E tests against Postgres, where it's more appropriate to be running in a full server environment.

> I never realized how complicated testing Flask is lol

... now you understand why I've put it off for 2 years.

jberkus commented:
> We could merge those for now, I feel like it would nice to get something new into the repo since it hasn't been touched for a few years.

Can you open a PR that's just the Schultze tests?

CheatCodeSam commented:
I've gone ahead and done so, please see #104

CheatCodeSam commented:
I've gone ahead and done so, please see #104
